src/index/silentpaymentindex.cpp

```diff
@@ -0,0 +1,134 @@
+{
+public:
+    explicit DB(size_t n_cache_size, bool f_memory = false, bool f_wipe = false);
+
+    bool WriteSilentPayments(const std::pair<uint256, std::vector<CPubKey>>& tweaks);
+};
+
+SilentPaymentIndex::DB::DB(size_t n_cache_size, bool f_memory, bool f_wipe) :
+    BaseIndex::DB(gArgs.GetDataDirNet() / "indexes" / "silentpaymentindex", n_cache_size, f_memory, f_wipe)
+{}
+
+bool SilentPaymentIndex::DB::WriteSilentPayments(const std::pair<uint256, std::vector<CPubKey>>& tweaks)
+{
+    CDBBatch batch(*this);
+    batch.Write(std::make_pair(DB_SILENT_PAYMENT_INDEX, tweaks.first), tweaks.second);
+    return WriteBatch(batch);
+}
+
+SilentPaymentIndex::SilentPaymentIndex(std::unique_ptr<interfaces::Chain> chain, size_t n_cache_size, bool f_memory, bool f_wipe)
+    : BaseIndex(std::move(chain), "silentpaymentindex"), m_db(std::make_unique<SilentPaymentIndex::DB>(n_cache_size, f_memory, f_wipe))
+{}
+
+SilentPaymentIndex::~SilentPaymentIndex() {}
+
+bool SilentPaymentIndex::GetSilentPaymentKeys(std::vector<CTransactionRef> txs, CBlockUndo& block_undo, std::vector<CPubKey>& tweaked_pub_key_sums)
+{
+    assert(txs.size() - 1 == block_undo.vtxundo.size());
+
+    for (uint32_t i=0; i < txs.size(); i++) {
+        auto& tx = txs.at(i);
+
+        if (tx->IsCoinBase()) {
+            continue;
+        }
+
+        for (auto& vout : tx->vout) {
+            if (std::none_of(tx->vout.begin(), tx->vout.end(), [vout](const CTxOut& txout) {
+                std::vector<std::vector<unsigned char>> solutions;
+                return Solver(vout.scriptPubKey, solutions) == TxoutType::WITNESS_V1_TAPROOT;
+            })) {
+                continue;
+            }
+        }
+
+        // -1 as blockundo does not have coinbase tx
+        CTxUndo undoTX{block_undo.vtxundo.at(i - 1)};
+        std::map<COutPoint, Coin> coins;
+
+        for (uint32_t j = 0; j < tx->vin.size(); j++) {
+            coins[tx->vin.at(j).prevout] = undoTX.vprevout.at(j);
+        }
+
+        const auto tweak_data = wallet::GetSilentPaymentsTweakDataFromTxInputs(tx->vin, coins);
+
+        if (!tweak_data) continue;
+
+        const uint256 outpoint_hash = tweak_data->first;
+        CPubKey tweaked_pub_key_sum = tweak_data->second;
+        if (!tweaked_pub_key_sum.TweakAdd(outpoint_hash.begin())) {
+            // TODO: Is this bad??
+            continue;
+        }
+        assert(tweaked_pub_key_sum.IsFullyValid());
+        tweaked_pub_key_sums.emplace_back(tweaked_pub_key_sum);
+    }
+
+    // TODO: so this can never fail?
+    return true;
+
+
+}
+
+bool SilentPaymentIndex::CustomAppend(const interfaces::BlockInfo& block)
+{
+    // Exclude genesis block transaction because outputs are not spendable.
+    if (block.height == 0) return true;
+
+    assert(block.data);
+
+    Consensus::Params consensus = Params().GetConsensus();
+
+    if (block.height < consensus.vDeployments[Consensus::DEPLOYMENT_TAPROOT].min_activation_height) {
```

Comment:
It might be nice if the indexer base class can skip ahead to a minimum height, as it takes a bunch of time to go through all the blocks.

Reply:
Yeah, we don't have the possibility to skip blocks yet, so I think we would need to add it as a feature on base index, I have started drafting it here, though it's not functional yet: https://github.com/fjahr/bitcoin/commit/cd7bf1249765a011210653c393b3a0fdb7106d55

But I also guess 2h doesn't sound extremely bad, we might want more than one silentpaymentindex to motivate something like this.